File: data/derived_cot/rq1_traces/translation/glm5.1/py_to_cpp/code_output/SQLQueryBuilder.cpp
#include <string>
#include <vector>
#include <utility>

class SQLQueryBuilder {
public:
    static std::string select(
        const std::string& table,
        const std::vector<std::string>& columns = {},
        const std::vector<std::pair<std::string, std::string>>& where = {})
    {
        std::string cols;
        if (columns.empty()) {
            cols = "*";
        } else {
            for (size_t i = 0; i < columns.size(); ++i) {
                if (i > 0) cols += ", ";
                cols += columns[i];
            }
        }
        std::string query = "SELECT " + cols + " FROM " + table;
        if (!where.empty()) {
            query += " WHERE ";
            for (size_t i = 0; i < where.size(); ++i) {
                if (i > 0) query += " AND ";
                query += where[i].first + "='" + where[i].second + "'";
            }
        }
        return query;
    }

    static std::string insert(
        const std::string& table,
        const std::vector<std::pair<std::string, std::string>>& data)
    {
        std::string keys, values;
        for (size_t i = 0; i < data.size(); ++i) {
            if (i > 0) {
                keys += ", ";
                values += ", ";
            }
            keys += data[i].first;
            values += "'" + data[i].second + "'";
        }
        return "INSERT INTO " + table + " (" + keys + ") VALUES (" + values + ")";
    }

    // Named delete_ because 'delete' is a C++ keyword
    static std::string delete_(
        const std::string& table,
        const std::vector<std::pair<std::string, std::string>>& where = {})
    {
        std::string query = "DELETE FROM " + table;
        if (!where.empty()) {
            query += " WHERE ";
            for (size_t i = 0; i < where.size(); ++i) {
                if (i > 0) query += " AND ";
                query += where[i].first + "='" + where[i].second + "'";
            }
        }
        return query;
    }

    static std::string update(
        const std::string& table,
        const std::vector<std::pair<std::string, std::string>>& data,
        const std::vector<std::pair<std::string, std::string>>& where = {})
    {
        std::string update_str;
        for (size_t i = 0; i < data.size(); ++i) {
            if (i > 0) update_str += ", ";
            update_str += data[i].first + "='" + data[i].second + "'";
        }
        std::string query = "UPDATE " + table + " SET " + update_str;
        if (!where.empty()) {
            query += " WHERE ";
            for (size_t i = 0; i < where.size(); ++i) {
                if (i > 0) query += " AND ";
                query += where[i].first + "='" + where[i].second + "'";
            }
        }
        return query;
    }
};

File: data/derived_cot/rq1_traces/translation/r1/java_to_py/code_output/SQLQueryBuilder.py
import re
from typing import Dict, List, Optional, Union

def select(table: str, columns: Union[str, List[str]] = "*", where: Optional[Dict[str, str]] = None) -> str:
    if columns is None:
        columns = "*"
    if columns == "*":
        columns_list = []
    elif isinstance(columns, str):
        columns_list = re.split(r',\s*', columns)
    elif isinstance(columns, (list, tuple)):
        columns_list = list(columns)
    else:
        raise TypeError("columns must be a string or a list/tuple of strings")
    
    if not columns_list:
        cols_str = "*"
    else:
        cols_str = ", ".join(columns_list)
    
    query = f"SELECT {cols_str} FROM {table}"
    
    if where is not None and where:
        conditions = []
        for col, val in where.items():
            conditions.append(f"{col}='{val}'")
        query += " WHERE " + " AND ".join(conditions)
    
    return query

def insert(table: str, data: Dict[str, str]) -> str:
    if not data:
        return f"INSERT INTO {table} () VALUES ()"
    
    cols = []
    vals = []
    for col, val in data.items():
        cols.append(col)
        vals.append(f"'{val}'")
    
    cols_str = ", ".join(cols)
    vals_str = ", ".join(vals)
    
    return f"INSERT INTO {table} ({cols_str}) VALUES ({vals_str})"

def delete(table: str, where: Optional[Dict[str, str]] = None) -> str:
    query = f"DELETE FROM {table}"
    
    if where is not None and where:
        conditions = []
        for col, val in where.items():
            conditions.append(f"{col}='{val}'")
        query += " WHERE " + " AND ".join(conditions)
    
    return query

def update(table: str, data: Dict[str, str], where: Optional[Dict[str, str]] = None) -> str:
    set_parts = []
    for col, val in data.items():
        set_parts.append(f"{col}='{val}'")
    set_clause = ", ".join(set_parts)
    
    query = f"UPDATE {table} SET {set_clause}"
    
    if where is not None and where:
        conditions = []
        for col, val in where.items():
            conditions.append(f"{col}='{val}'")
        query += " WHERE " + " AND ".join(conditions)
    
    return query